### PR TITLE
provide ratings and feedback

### PR DESCRIPTION
**Pull request by Fidela1**
### What does this PR do?

- Provide feedback and ratings

### Description of Task to be completed?

- Provide feedback and ratings for product bought

### How should this be manually tested?

1. clone the [repo](url)
2. Checkout to this branch `feature/provide-ratings-184347035`
3. Run `npm run dev`

### What are the relevant pivotal tracker stories?
[184347035](https://www.pivotaltracker.com/story/show/184347035)

### CODE REVIEW

**Review by irakozetony — APPROVED** [review]

**Review by irakozetony — APPROVED** [review]

**Review by ndatumuremyi — APPROVED** [review]

File: src/middlewares/product.middleware.js [hunk 1]
```diff
@@ -73,8 +73,10 @@ export const checkIfProductExistsById = async (req, res, next) => {
 };
 
 export const checkIfSellerOwnsProduct = async (req, res, next) => {
+  const productId =req.sale?.productId || req.params.productId;
+  console.log(productId)
```
Comment (ncutixavier) [inline src/middlewares/product.middleware.js]: remove all logs

File: src/middlewares/ratings.middleware.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import { OrderService } from '../services/order.service';
+
+export const validateRating = async (req, res, next) => {
+    const rate = req.body.rate;
+    if (rate <0 || rate >5 ){
+       return res.status(400).json({
+      message: 'Rate must be from 0 to 5',});
```
Comment (ncutixavier) [inline src/middlewares/ratings.middleware.js]: Ranged from 1-5 not from 0

File: src/controllers/ratings.controller.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      productId,
+    };
+    let rate = await RatingService.getProductRatingByField({ buyerId: buyerId });
+
+    if (rate) {
+      await RatingService.updateRatings(ratings, rate.id);
+      rate = await RatingService.getProductRatingByField({ id: rate.id });
+    }
+    else {
+      rate = await RatingService.productRating(ratings);
```
Comment (ncutixavier) [inline src/controllers/ratings.controller.js]: Rename `productRating` to `createRating`

File: src/controllers/ratings.controller.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  } catch (err) {
+    return res.status(500).json({
+      error: err.message,
+      message: 'Error occured while rating a product',
+    });
+  }
+
+};
+
+export const getRatingsById = async (req, res) => {
```
Comment (ncutixavier) [inline src/controllers/ratings.controller.js]: Rename this controller to be specific for getting ratings by specific product

File: src/middlewares/ratings.middleware.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+  export const checkIfUserBoughtProduct = async (req, res, next) => {
+    const productId = req.body.productId;
+    const buyerId = req.user.id;
+    const orders = await OrderService.getOrderByBuyerIdAndProductId(buyerId, productId);
+    if(!orders || !orders.length){
+        return res.status(401).json({
+             message:'You have to buy this product first in order to provide ratings and feedback', });
+           }
+      next();
```
Comment (ncutixavier) [inline src/middlewares/ratings.middleware.js]: How do you know if the order has completed? because I can't see any condition

**Review by istafrica-dennw — CHANGES_REQUESTED** [review]
Your endpoint is wrong

**Comment by istafrica-dennw** [comment]
> Your endpoint is wrong

I gave you feedback on the document but you never worked on them, so I will wait for you to fix this

File: src/routes/api/rating.route.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  checkIfUserBoughtProduct,
+  productRating
+);
+
+ratingRoute.get('/:productId',
+protectRoute,
+uuidValidation('productId'),
+checkIfSellerOwnsProduct,
+getRatingsById
+);
```
Comment (ncutixavier) [inline src/routes/api/rating.route.js]: - Change the endpoint as I showed you yesterday
- Do you know why you check if the seller owns a product? multiple users can view the product ratings. You don't need to restrict them

**Review by ncutixavier — CHANGES_REQUESTED** [review]

File: src/controllers/ratings.controller.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+};
+
+export const getRatingsBySpecificProduct = async (req, res) => {
+  try {
+    const productId = req.params.productId;
+
+    const ratings = await RatingService.getProductRatingByField({productId: productId});
+    return res.status(201).json({ message: 'Ratings retrived successfully', ratings: ratings });
```
Comment (ncutixavier) [inline src/controllers/ratings.controller.js]: Update status code

File: src/middlewares/ratings.middleware.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import { OrderService } from '../services/order.service';
+
+export const validateRating = async (req, res, next) => {
+  const rate = req.body.rate;
+  if (rate < 1 || rate > 5) {
+    return res.status(400).json({
+      message: 'Rate must be from 0 to 5',
+    });
+  }
+  next();
```
Comment (ncutixavier) [inline src/middlewares/ratings.middleware.js]: Rate should be ranged from 1-5

**Review by ncutixavier — CHANGES_REQUESTED** [review]

File: src/middlewares/ratings.middleware.js [hunk 9]
```diff
@@ -0,0 +1,23 @@
+import { OrderService } from '../services/order.service';
+
+export const validateRating = async (req, res, next) => {
+  const rate = req.body.rate;
+  if (rate < 1 || rate > 5) {
```
Comment (ncutixavier) [inline src/middlewares/ratings.middleware.js]: 1 and 5 should be included

**Review by ncutixavier — APPROVED** [review]

**Pull request merged**